pkg/controller/nodelifecycle/node_lifecycle_controller_test.go

```diff
@@ -3057,3 +3057,210 @@ func TestReconcileNodeLabels(t *testing.T) {
+						{
+							Type:               v1.NodeReady,
+							Status:             v1.ConditionUnknown,
+							LastHeartbeatTime:  fakeNow,
+							LastTransitionTime: fakeNow,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "Status nil",
+			node: &v1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:              "node0",
+					CreationTimestamp: fakeNow,
+				},
+				Status: v1.NodeStatus{
+					Conditions: []v1.NodeCondition{},
+				},
+			},
+		},
+		{
+			name: "Status true - after grace period",
+			node: &v1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:              "node0",
+					CreationTimestamp: fakeOld,
+				},
+				Status: v1.NodeStatus{
+					Conditions: []v1.NodeCondition{
+						{
+							Type:               v1.NodeReady,
+							Status:             v1.ConditionTrue,
+							LastHeartbeatTime:  fakeOld,
+							LastTransitionTime: fakeOld,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "Status false - after grace period",
+			node: &v1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:              "node0",
+					CreationTimestamp: fakeOld,
+				},
+				Status: v1.NodeStatus{
+					Conditions: []v1.NodeCondition{
+						{
+							Type:               v1.NodeReady,
+							Status:             v1.ConditionFalse,
+							LastHeartbeatTime:  fakeOld,
+							LastTransitionTime: fakeOld,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "Status unknown - after grace period",
+			node: &v1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:              "node0",
+					CreationTimestamp: fakeOld,
+				},
+				Status: v1.NodeStatus{
+					Conditions: []v1.NodeCondition{
+						{
+							Type:               v1.NodeReady,
+							Status:             v1.ConditionUnknown,
+							LastHeartbeatTime:  fakeOld,
+							LastTransitionTime: fakeOld,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "Status nil - after grace period",
+			node: &v1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:              "node0",
+					CreationTimestamp: fakeOld,
+				},
+				Status: v1.NodeStatus{
+					Conditions: []v1.NodeCondition{},
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		nodeController.nodeHealthMap[test.node.Name] = &nodeHealthData{
+			status:                   &test.node.Status,
+			probeTimestamp:           test.node.CreationTimestamp,
+			readyTransitionTimestamp: test.node.CreationTimestamp,
+		}
+		_, _, currentReadyCondition, err := nodeController.tryUpdateNodeHealth(test.node)
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+		_, savedReadyCondition := nodeutil.GetNodeCondition(nodeController.nodeHealthMap[test.node.Name].status, v1.NodeReady)
+		if (currentReadyCondition == nil) != (savedReadyCondition == nil) {
+			t.Errorf("%s: expected %v, got %v", test.name, savedReadyCondition, currentReadyCondition)
+		}
+
+		if (currentReadyCondition != nil && savedReadyCondition != nil) &&
+			currentReadyCondition.Status != savedReadyCondition.Status {
```

Comment:
add a helper function:

getStatus(cond) {
  if cond == nil {
    return nil
  }
  return cond.Status
}

With that, all you need to check is:
if getStatus(currentReadyCond) != getStatus(savedReadyCond) {
  t.Errorf()
}